src/coreclr/jit/morph.cpp

```diff
@@ -14135,16 +14145,17 @@ void Compiler::fgMergeBlockReturn(BasicBlock* block)
                 lastStmt = newStmt;
             }
         }
-        else if (ret != nullptr && ret->OperGet() == GT_RETURN)
+        else if ((ret != nullptr) && ret->OperIs(GT_RETURN, GT_SWIFT_ERROR_RET))
         {
             // This block ends with a GT_RETURN
             noway_assert(lastStmt != nullptr);
             noway_assert(lastStmt->GetNextStmt() == nullptr);
 
-            // Must be a void GT_RETURN with null operand; delete it as this block branches to oneReturn
-            // block
+            // Must be a void GT_RETURN/GT_SWIFT_ERROR_RET with null operand; delete it as this block branches to
+            // oneReturn block
+            GenTree* const retVal = ret->OperIs(GT_RETURN) ? ret->gtGetOp1() : ret->gtGetOp2();
             noway_assert(ret->TypeGet() == TYP_VOID);
-            noway_assert(ret->gtGetOp1() == nullptr);
+            noway_assert(retVal == nullptr);
```

Comment:
I think this logic isn't quite sufficient since it runs as part of morph and there are transformations between import and morph. For example, try the following example:
```csharp
[UnmanagedCalersOnly(...Swift...)]
int Foo(SwiftError* err)
{
  if (Environment.TickCount % 2 == 0)
  {
    *err = GetError(1);
    return 10;
  }
  else
  {
    *err = GetError(2);
    return 20;
  }
}
```

I would expect forward sub to come along and transform
```
GT_STORE_LCL_VAR(error local, GetError)
GT_SWIFT_ERROR_RET(error local, 10)
```
into
```
GT_SWIFT_ERROR_RET(GetError(), 10)
```
and similar for the other return. Then this return merging will silently remove the call to `GetError` in favor of "error local" that no longer has any definitions.

I didn't test the example, so we may hit some stops underway that makes it work out anyway, but regardless I think we should make this more robust by introducing similar stores to the error return local based on what the argument to `GT_SWIFT_ERROR_RET` is (can be skipped when the operand is already the local, of course, which is probably common).

Reply:
> can be skipped when the operand is already the local, of course, which is probably common

Forward substitution doesn't kick in with the test you shared, but I noticed that struct promotion replaces all the `GT_LCL_FIELD` usages of `SwiftError::Value` with `GT_LCL_VAR` usages of the `Value` field, which complicates detecting if the operand is the error local because the `lclNums` are no longer what we expect; I'm guessing we don't want to negate the benefits of this promotion by always storing the promoted local back to the struct. Do you think we'd be better off with just ensuring we don't do forward substitution of the error value?

For what it's worth, I think we're sufficiently blocking forward substitution from happening for now, because either we don't do struct promotion for the pseudolocal and the `GT_SWIFT_ERROR_RET`'s error operand is a `GT_LCL_FLD` (correct me if I'm wrong, but I don't think forward substitution tries to handle cases like this), or we do struct promotion and end up with a pattern like this:
```
------------ BB03 [0002] [011..01E) (return), preds={BB01} succs={}

***** BB03 [0002]
STMT00001 ( 0x011[E-] ... 0x018 )
               [000007] DAC-G------                         *  STORE_LCL_VAR struct<System.Runtime.InteropServices.Swift.SwiftError, 8>(P) V02 tmp0         
                                                            *    long   field V02.<Value>k__BackingField (fldOffset=0x0) -> V05 tmp3         
               [000006] --C-G------                         \--*  CALL      struct ErrorHandlingTests:GetError(int):System.Runtime.InteropServices.Swift.SwiftError
               [000005] ----------- arg0                       \--*  CNS_INT   int    0

***** BB03 [0002]
STMT00002 ( 0x01D[E-] ... 0x01D )
               [000008] -----------                         *  SWIFT_ERROR_RET void  
               [000009] -----------                         \--*  LCL_VAR   long   V05 tmp3          (last use)

------------ BB04 [0003] [???..???) (return), preds={} succs={}

***** BB04 [0003]
STMT00007 ( ??? ... ??? )
               [000021] --C-G------                         *  CALL help void   CORINFO_HELP_JIT_REVERSE_PINVOKE_EXIT
               [000020] ----------- arg0                    \--*  LCL_ADDR  long   V04 tmp2         [+0]

***** BB04 [0003]
STMT00005 ( ??? ... ??? )
               [000016] -----------                         *  SWIFT_ERROR_RET void  
               [000017] -----------                         \--*  LCL_VAR   long   V05 tmp3          (last use)
```

Struct promotion seems to block that `GT_STORE_LCL_VAR` from being substituted:
```
*************** Starting PHASE Forward Substitution
...
===> BB03
    [000007]:  next stmt has non-last use
```

Reply:
> Forward substitution doesn't kick in with the test you shared, but I noticed that struct promotion replaces all the GT_LCL_FLD usages of SwiftError::Value with GT_LCL_VAR usages of the Value field, which complicates detecting if the operand is the error local because the lclNums are no longer what we expect; I'm guessing we don't want to negate the benefits of this promotion by always storing the promoted local back to the struct. Do you think we'd be better off with just ensuring we don't do forward substitution of the error value?

We should not need to change forward sub nor promotion here. We just need to make sure that return merging handles this extra operand to `GT_SWIFT_ERROR_RET` correctly, and also merges the error for these nodes, in the same way that the return values are merged. It needs to introduce stores back to the pseudolocal from the error operand to `GT_SWIFT_ERROR_RET`.

> For what it's worth, I think we're sufficiently blocking forward substitution from happening for now, because either we don't do struct promotion for the pseudolocal and the GT_SWIFT_ERROR_RET's error operand is a GT_LCL_FLD (correct me if I'm wrong, but I don't think forward substitution tries to handle cases like this), or we do struct promotion and end up with a pattern like this:

With old promotion forward sub is not able to do much for these kinds of patterns because we handle a large part of the promotion transformation during morph that runs after forward sub. But with new promotion you can definitely see the highlighted issue. It doesn't happen for calls because of how we represent calls, but try this one:
```csharp
[UnmanagedCallersOnly(CallConvs = [typeof(CallConvSwift)])]
private static int Foo(SwiftError* err, int* errorValue)
{
    if (Environment.TickCount % 2 == 0)
    {
        *err = new SwiftError((void*)(*errorValue + 1));
        return 10;
    }
    else
    {
        *err = new SwiftError((void*)(*errorValue + 2));
        return 20;
    }
}
```
Disabling old promotion with `DOTNET_JitStressModeNames=STRESS_NO_OLD_PROMOTION` (there are also natural case where old promotion turns off and new promotion would kick in) results in the following IR after physical promotion:
```scala
------------ BB05 [0006] [???..???) -> BB01(1) (always), preds={} succs={BB01}

***** BB05 [0006]
STMT00014 ( ??? ... ??? )
               [000055] --C-G------                         ▌  CALL help void   CORINFO_HELP_JIT_REVERSE_PINVOKE_ENTER
               [000054] ----------- arg0                    └──▌  LCL_ADDR  long   V09 tmp7         [+0]

------------ BB01 [0000] [000..009) -> BB03(0.5),BB02(0.5) (cond), preds={BB05} succs={BB02,BB03}

***** BB01 [0000]
STMT00000 ( 0x000[E-] ... 0x007 )
               [000005] --C-G------                         ▌  JTRUE     void  
               [000004] --C-G------                         └──▌  NE        int   
               [000002] --C-G------                            ├──▌  MOD       int   
               [000000] --C-G------                            │  ├──▌  CALL      int    System.Environment:get_TickCount():int
               [000001] -----------                            │  └──▌  CNS_INT   int    2
               [000003] -----------                            └──▌  CNS_INT   int    0

------------ BB02 [0001] [009..01C) (return), preds={BB01} succs={}

***** BB02 [0001]
STMT00008 ( 0x019[E-] ... 0x01B )
               [000037] ---XG------                         ▌  SWIFT_ERROR_RET int   
               [000027] ---XG------                         ├──▌  CAST      long <- int
               [000026] ---XG------                         │  └──▌  ADD       int   
               [000024] ---XG------                         │     ├──▌  IND       int   
               [000023] -----------                         │     │  └──▌  LCL_VAR   long   V01 arg1          (last use)
               [000025] -----------                         │     └──▌  CNS_INT   int    1
               [000034] -----------                         └──▌  CNS_INT   int    10

------------ BB03 [0002] [01C..02F) (return), preds={BB01} succs={}

***** BB03 [0002]
STMT00004 ( 0x02C[E-] ... 0x02E )
               [000021] ---XG------                         ▌  SWIFT_ERROR_RET int   
               [000011] ---XG------                         ├──▌  CAST      long <- int
               [000010] ---XG------                         │  └──▌  ADD       int   
               [000008] ---XG------                         │     ├──▌  IND       int   
               [000007] -----------                         │     │  └──▌  LCL_VAR   long   V01 arg1          (last use)
               [000009] -----------                         │     └──▌  CNS_INT   int    2
               [000018] -----------                         └──▌  CNS_INT   int    20

------------ BB04 [0005] [???..???) (return), preds={} succs={}

***** BB04 [0005]
STMT00015 ( ??? ... ??? )
               [000057] --C-G------                         ▌  CALL help void   CORINFO_HELP_JIT_REVERSE_PINVOKE_EXIT
               [000056] ----------- arg0                    └──▌  LCL_ADDR  long   V09 tmp7         [+0]

***** BB04 [0005]
STMT00013 ( ??? ... ??? )
               [000053] -----------                         ▌  SWIFT_ERROR_RET int   
               [000066] -----------                         ├──▌  LCL_VAR   long   V10 tmp8          (last use)
               [000050] -------N---                         └──▌  LCL_VAR   int    V08 tmp6          (last use)
```

After morph we have:
```scala
------------ BB05 [0006] [???..???) -> BB01(1) (always), preds={} succs={BB01}

***** BB05 [0006]
STMT00014 ( ??? ... ??? )
               [000055] --C-G+-----                         ▌  CALL help void   CORINFO_HELP_JIT_REVERSE_PINVOKE_ENTER
               [000054] -----+----- arg0 in rdi             └──▌  LCL_ADDR  long   V09 tmp7         [+0]

------------ BB01 [0000] [000..009) -> BB03(0.5),BB02(0.5) (cond), preds={BB05} succs={BB02,BB03}

***** BB01 [0000]
STMT00000 ( 0x000[E-] ... 0x007 )
               [000005] --CXG+-----                         ▌  JTRUE     void  
               [000004] J-CXG+-N---                         └──▌  NE        int   
               [000002] --CXG+-----                            ├──▌  AND       int   
               [000000] --CXG+-----                            │  ├──▌  CALL      int    System.Environment:get_TickCount():int
               [000001] -----+-----                            │  └──▌  CNS_INT   int    1
               [000003] -----+-----                            └──▌  CNS_INT   int    0

------------ BB02 [0001] [009..01C) -> BB04(1) (always), preds={BB01} succs={BB04}

***** BB02 [0001]
STMT00008 ( 0x019[E-] ... 0x01B )
               [000070] DA---------                         ▌  STORE_LCL_VAR int    V08 tmp6         
               [000034] -----+-----                         └──▌  CNS_INT   int    10

------------ BB03 [0002] [01C..02F) -> BB04(1) (always), preds={BB01} succs={BB04}

***** BB03 [0002]
STMT00004 ( 0x02C[E-] ... 0x02E )
               [000069] DA---------                         ▌  STORE_LCL_VAR int    V08 tmp6         
               [000018] -----+-----                         └──▌  CNS_INT   int    20

------------ BB04 [0005] [???..???) (return), preds={BB02,BB03} succs={}

***** BB04 [0005]
STMT00015 ( ??? ... ??? )
               [000057] --C-G+-----                         ▌  CALL help void   CORINFO_HELP_JIT_REVERSE_PINVOKE_EXIT
               [000056] -----+----- arg0 in rdi             └──▌  LCL_ADDR  long   V09 tmp7         [+0]

***** BB04 [0005]
STMT00013 ( ??? ... ??? )
               [000053] -----+-----                         ▌  SWIFT_ERROR_RET int   
               [000066] -----+-----                         ├──▌  LCL_VAR   long   V10 tmp8          (last use)
               [000050] -----+-N---                         └──▌  LCL_VAR   int    V08 tmp6          (last use)
```

where we've lost the error return values. We need to introduce the similar stores as `[000070]` and `[000069]`.

It's a bit of a blemish that the return merging is split across `fgAddInternal` and morph like this. It would be a nice clean up to do all of this early. I think @AndyAyersMS was trying to do that at some point, but hit some issues around it.

Reply:
I see, thanks for the explanation! I think I was overthinking what this store should look like. Just to check:

* If struct promotion is disabled (or we just don't promote the pseudolocal, for whatever reason), and we write directly to `SwiftError::Value`, we will access the error value with a `GT_LCL_FLD`, and any stores to it will use `GT_STORE_LCL_FLD`. Forward substitution only works on `GT_STORE_LCL_VAR`, so during return merging, we can just check if the error operand is a `GT_LCL_FLD`, and if so, skip the store.
* If we promote the pseudolocal, we will use `GT_LCL_VAR`/`GT_STORE_LCL_VAR` to read/write the promoted error local, and we can do forward substitution on such stores. During return merging, we can check if the pseudolocal was promoted, and get the correct `lclNum` of the error local; then, `gtNewTempStore` will take care of the rest.
* If we don't promote the pseudolocal, and then set the whole struct with `GT_STORE_LCL_VAR` (I'm guessing this is possible with patterns like `*err = GetError(1)`?), `gtNewTempStore` should handle this too if we give it the `lclNum` of the pseudolocal struct (which is `Compiler::lvaSwiftErrorLocal`).

Reply:
I don't think it needs to be that complicated. I was getting a bit confused around the pseudo-local, thinking it was a `TYP_I_IMPL` typed local. But it's a struct. So we should just introduce the counterpart to `genReturnLocal` that is a normal `TYP_I_IMPL` local and just always create the stores to it in morph when doing return merging. The potential extra stores should not be a problem, they can be removed by copy prop or LSRA can make them zero cost by preferencing them to the same register, which should happen automatically.